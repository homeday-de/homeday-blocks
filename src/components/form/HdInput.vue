<template>
  <div
    :class="fieldClasses"
    class="field field--input"
  >
    <img
      v-if="icon"
      :src="icon"
      role="presentation"
      class="field__icon"
    >
    <input
      :autocomplete="autocomplete"
      :value="value"
      :type="currentType"
      :id="name"
      :name="name"
      :placeholder="isActive && placeholder !== undefined ? placeholder : ''"
      :required="required"
      :autofocus="autofocus"
      :disabled="disabled"
      class="field__input"
      ref="input"
      data-lpignore="true"
      @input="handleInput"
      @focus="handleFocus"
      @blur="handleBlur"
    />
    <label
      v-if="label"
      :for="name"
      class="field__label"
    >
      {{ label }}
    </label>
    <p
      v-if="error"
      class="field__error"
    >
      {{ error }}
    </p>
    <p
      v-else-if="helper"
      class="field__error field__error--helper"
      v-html="helper"
    />
    <span class="field__border"/>
  </div>
</template>

<script>
import merge from 'lodash/merge';
import { getMessages } from 'homeday-blocks/src/lang';
import {
  email as validateEmail,
  date as validateDate,
} from 'homeday-blocks/src/services/formValidation';

export default {
  name: 'HdInput',
  props: {
    label: {
      type: String,
      default: '',
    },
    name: {
      type: String,
      required: true,
    },
    type: {
      type: String,
      default: 'text',
    },
    value: {
      type: [String, Number],
      default: '',
    },
    placeholder: {
      type: String,
      default: '',
    },
    required: {
      type: Boolean,
      default: false,
    },
    autocomplete: {
      type: String,
      default: 'on',
    },
    autofocus: {
      type: Boolean,
      default: false,
    },
    min: {
      type: Number,
      default: undefined,
    },
    max: {
      type: Number,
      default: undefined,
    },
    lang: {
      type: String,
      default: 'de',
    },
    texts: {
      type: Object,
      default: () => ({}),
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    icon: {
      type: String,
      default: '',
    },
    customRules: {
      type: Array,
      default: () => [],
      validator: rulesProvided => rulesProvided.every(
        ({ validate, errorMessage }) => typeof validate === 'function' && typeof errorMessage === 'string',
      ),
    },
  },
  data() {
    return {
      currentType: this.type || 'text',
      isActive: undefined,
      isValid: undefined,
      error: null,
      helper: null,
    };
  },
  computed: {
    t() {
      return merge(getMessages(this.lang), this.texts);
    },
    isEmpty() {
      return this.value === null || this.value === undefined || this.value === '';
    },
    fieldClasses() {
      return {
        'field--active': this.isActive,
        'field--filled': !this.isEmpty,
        'field--invalid': this.isValid === false,
        'field--disabled': this.disabled,
        'field--hasIcon': this.icon,
      };
    },
  },
  watch: {
    value() {
      this.validate();
    },
    type(type) {
      this.currentType = type;
    },
  },
  methods: {
    clearInput() {
      this.$refs.input.focus();
      this.hideError();
      this.$emit('input', '');
    },
    handleFocus() {
      this.isActive = true;
      this.$emit('focus');
    },
    handleBlur() {
      this.isActive = false;
      this.validate();
      this.$emit('blur');
    },
    handleInput(e) {
      let newValue = e.target.value;

      if (this.currentType === 'number' && newValue !== '') {
        newValue = parseFloat(newValue);

        if (typeof this.min === 'number' && newValue < this.min) {
          newValue = this.min;
        } else if (typeof this.max === 'number' && newValue > this.max) {
          newValue = this.max;
        }

        if (newValue === this.value) {
          this.$nextTick(this.$forceUpdate);
        }
      }

      this.$emit('input', newValue);
    },
    showError(errorMessage) {
      this.error = errorMessage;
      this.isValid = false;
    },
    showHelper(message) {
      this.helper = message;
    },
    hideError() {
      this.isValid = true;
      this.error = null;
    },
    validate() {
      if (this.required && this.isEmpty) {
        this.showError(this.t.FORM.VALIDATION.REQUIRED);
      } else if (this.currentType === 'email' && !validateEmail(this.value)) {
        this.showError(this.t.FORM.VALIDATION.INVALID_EMAIL);
      } else if (this.currentType === 'date' && !validateDate(this.value)) {
        this.showError(this.t.FORM.VALIDATION.INVALID_DATE);
      } else if (this.customRules.length && !this.isEmpty) {
        this.validateCustomRules();
      } else {
        this.hideError();
      }

      return !this.error;
    },
    validateCustomRules() {
      const firstFailingRule = this.customRules.find(({ validate }) => !validate(this.value));
      if (firstFailingRule) {
        this.showError(firstFailingRule.errorMessage);
      } else {
        this.hideError();
      }
    },
  },
};
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped lang="scss">
@import 'homeday-blocks/src/styles/mixins.scss';
@import 'homeday-blocks/src/styles/inputs.scss';

.field {
  &__error {
    width: 100%;
    text-align: left;
    &--helper {
      display: block;
      color: getShade($quaternary-color, 80);
    }
  }
}
</style>
