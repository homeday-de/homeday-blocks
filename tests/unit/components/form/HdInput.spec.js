import { wrapperFactoryBuilder } from 'tests/unit/helpers';
import FIELD_CLASSES from './FIELD_CLASSES';
import HdInput from '@/components/form/HdInput.vue';

const ERROR_SELECTOR = '.field__error';
const HELPER_SELECTOR = '.field__error--helper';
const ICON_SELECTOR = '.field__icon';

const TEST_VALUE = 'new value';
const TEST_PROPS = {
  name: 'test name',
  label: 'test label',
  placeholder: 'test placeholder',
};

describe('HdInput', () => {
  const wrapperFactory = wrapperFactoryBuilder(
    HdInput,
    {
      propsData: TEST_PROPS,
    },
  );

  let wrapper;
  beforeEach(() => {
    wrapper = wrapperFactory();
  });

  test('is rendered as expected', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('is rendered with prefilled value', () => {
    wrapper = wrapperFactory({
      propsData: {
        value: TEST_VALUE,
      },
    });
    expect(wrapper.find('input').element.value).toBe(TEST_VALUE);
  });

  test('updates the field on change of the prop `value`', () => {
    wrapper.setProps({ value: TEST_VALUE });
    expect(wrapper.find('input').element.value).toBe(TEST_VALUE);
  });

  test('emits `input` event with the right payload', () => {
    wrapper.find('input').setValue(TEST_VALUE);
    expect(wrapper.emitted('input')[0][0]).toEqual(TEST_VALUE);
  });

  test('allows setting a helper', () => {
    const TEST_HTML = '<b>test</b> test';
    wrapper.vm.showHelper(TEST_HTML);
    expect(wrapper.find(HELPER_SELECTOR).element.innerHTML).toEqual(TEST_HTML);
  });

  test('validates requiredness', () => {
    wrapper = wrapperFactory({
      propsData: {
        value: '',
        required: true,
      },
    });
    expect(wrapper.vm.validate()).toBe(false);
    expect(wrapper.find(ERROR_SELECTOR).text()).toBeTruthy();
    expect(wrapper.classes()).toContain(FIELD_CLASSES.INVALID);

    wrapper.setProps({ required: false });
    expect(wrapper.vm.validate()).toBe(true);
  });

  describe('type "email"', () => {
    beforeEach(() => {
      wrapper = wrapperFactory({
        propsData: {
          type: 'email',
        },
      });
    });

    test('sets the input type to "email"', () => {
      expect(wrapper
        .find('input')
        .attributes()
        .type).toBe('email');
    });

    test('validates email', () => {
      const VALID_EMAIL = 'valid@email.com';
      const INVALID_EMAIL = 'unvalid@email@com';

      wrapper.setProps({
        type: 'email',
        value: INVALID_EMAIL,
      });

      expect(wrapper.vm.validate()).toBe(false);
      expect(wrapper.find(ERROR_SELECTOR).text()).toBeTruthy();
      expect(wrapper.classes()).toContain(FIELD_CLASSES.INVALID);

      wrapper.setProps({ value: VALID_EMAIL });
      expect(wrapper.vm.validate()).toBe(true);
    });
  });

  describe('type "date"', () => {
    beforeEach(() => {
      wrapper = wrapperFactory({
        propsData: {
          type: 'date',
        },
      });
    });

    test('sets the input type to "date"', () => {
      expect(wrapper
        .find('input')
        .attributes()
        .type).toBe('date');
    });

    test('validates date', () => {
      const VALID_DATE = '1990-01-01';
      const INVALID_DATE = '01-01-1990';

      wrapper.setProps({
        type: 'date',
        value: INVALID_DATE,
      });

      expect(wrapper.vm.validate()).toBe(false);
      expect(wrapper.find(ERROR_SELECTOR).text()).toBeTruthy();
      expect(wrapper.classes()).toContain(FIELD_CLASSES.INVALID);

      wrapper.setProps({ value: VALID_DATE });
      expect(wrapper.vm.validate()).toBe(true);
    });
  });

  describe('type "number"', () => {
    beforeEach(() => {
      wrapper = wrapperFactory({
        propsData: {
          type: 'number',
        },
      });
    });

    test('sets the input type to "number"', () => {
      expect(wrapper
        .find('input')
        .attributes()
        .type).toBe('number');
    });

    test('parses to Number', () => {
      const TEST_STRING = '1234.56';

      const input = wrapper.find('input');
      input.setValue(TEST_STRING);

      expect(wrapper.emitted('input')[0][0]).toBe(parseFloat(TEST_STRING));
    });

    test('limits the value in a given "min"-"max" range', () => {
      const MAX = 100;
      const MIN = 0;
      const TEST_NUMBER = 50;

      wrapper.setProps({
        max: MAX,
        min: MIN,
      });

      const input = wrapper.find('input');

      input.setValue(TEST_NUMBER);
      expect(wrapper.emitted('input')[0][0]).toBe(TEST_NUMBER);

      input.setValue(MAX + 10);
      expect(wrapper.emitted('input')[1][0]).toBe(MAX);

      input.setValue(MIN - 10);
      expect(wrapper.emitted('input')[2][0]).toBe(MIN);
    });
  });

  test('Supports disabling', () => {
    wrapper.setProps({
      disabled: true,
    });
    expect(wrapper.find('input').attributes().disabled).toBe('disabled');
  });

  test('Can render an icon', () => {
    const ICON_PATH = 'fake/icon.svg';

    expect(wrapper.classes()).not.toContain(FIELD_CLASSES.HAS_ICON);
    expect(wrapper.find(ICON_SELECTOR).exists()).toBe(false);

    wrapper.setProps({
      icon: ICON_PATH,
    });

    expect(wrapper.classes()).toContain(FIELD_CLASSES.HAS_ICON);
    expect(wrapper.find(ICON_SELECTOR).exists()).toBe(true);
  });

  test('ignores LastPass auto fill', () => {
    expect(wrapper
      .find('input')
      .attributes()['data-lpignore'])
      .toBe('true');
  });

  test('custom validation prop validation', () => {
    expect(wrapper.vm.$options.props.customRules.validator([1])).toBeFalsy();
    expect(wrapper.vm.$options.props.customRules.validator([{ validate: () => true, errorMessage: '' }])).toBeTruthy();
  });

  test('custom validation error message', async () => {
    // Set a custom validation where input should match 'custom' keyword
    const errorMsg = 'Input should match custom keyword';
    const customKeyword = 'custom';
    wrapper.setProps({
      customRules: [{
        validate: value => value === 'custom',
        errorMessage: errorMsg,
      }],
      value: customKeyword,
    });

    // Makes sure it does not display error msg
    expect(wrapper.vm.$data.error).toBeNull();

    // Change value and look for error message
    wrapper.setData({
      value: TEST_VALUE,
    });
    expect(wrapper.find(ERROR_SELECTOR).text()).toBe(errorMsg);
  });
});
