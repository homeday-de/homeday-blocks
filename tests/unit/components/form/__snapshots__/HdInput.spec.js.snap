// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HdInput is rendered as expected 1`] = `
<div class="field field--input">
  <!----> <input autocomplete="on" type="text" id="test name" name="test name" placeholder="" data-lpignore="true" class="field__input"> <label for="test name" class="field__label">
    test label
  </label>
  <!----> <span class="field__border"></span></div>
`;
